core/src/main/scala/filodb.core/downsample/DownsampleConfig.scala

```diff
@@ -18,6 +19,77 @@ final case class DownsampleConfig(config: Config) {
   val resolutions = if (config.hasPath ("resolutions")) config.as[Seq[FiniteDuration]]("resolutions")
                     else Seq.empty
 
+  /**
+   * DFS through nested mutable Maps, and return an equivalent tree of nested immutable Maps.
+   */
+  private def mapToImmutable(map: mutable.Map[String, Any]): Map[String, Any] = {
+    if (map.isEmpty) {
+      return Map()
+    }
+    map.map{ case (k, v) =>
+      k -> mapToImmutable(v.asInstanceOf[mutable.Map[String, Any]])
+    }.toMap
+  }
+
+  /**
+   * Convert data-shape keys into a map, where exactly the set of unique keys is described
+   *   by the set of unique paths through the maps.
+   */
+  private def dataShapeKeysToMap(keys: Seq[Seq[String]]): Map[String, Any] = {
```

Comment:
are you building a trie tree? If so, maybe naming with trie terminology will help understanding.

Reply:
That's exactly right-- I will update the terminology used here and elsewhere.

Reply:
No need to build mutable and then make it immutable, its possible to build the immutable map in the first place